collect_app/src/main/java/org/odk/collect/android/listeners/SwipeHandler.java

```diff
@@ -0,0 +1,152 @@
+import org.odk.collect.android.formentry.ODKView;
+import org.odk.collect.android.preferences.GeneralKeys;
+import org.odk.collect.android.preferences.GeneralSharedPreferences;
+import org.odk.collect.android.utilities.FlingRegister;
+import org.odk.collect.android.utilities.ScreenUtils;
+
+import timber.log.Timber;
+
+public class SwipeHandler {
+    private final GestureDetector gestureDetector;
+    private final OnSwipeListener onSwipe;
+    private ODKView odkView;
+    private boolean allowSwiping = true;
+    private boolean beenSwiped;
+
+    public interface OnSwipeListener {
+        void showPreviousView();
+        void showNextView();
+    }
+
+    public SwipeHandler(Context context) {
+        gestureDetector = new GestureDetector(context, new GestureListener());
+        this.onSwipe = (OnSwipeListener) context;
+    }
+
+    public void setOdkView(ODKView odkView) {
+        this.odkView = odkView;
+    }
+
+    public void setAllowSwiping(boolean allowSwiping) {
+        this.allowSwiping = allowSwiping;
+    }
+
+    public void setBeenSwiped(boolean beenSwiped) {
+        this.beenSwiped = beenSwiped;
+    }
+
+    public boolean beenSwiped() {
+        return beenSwiped;
+    }
+
+    public GestureDetector getGestureDetector() {
+        return gestureDetector;
+    }
+
+    public class GestureListener implements GestureDetector.OnGestureListener {
+
+        @Override
+        public boolean onDown(MotionEvent event) {
+            return false;
+        }
+
+        @Override
+        public void onShowPress(MotionEvent e) {
+
+        }
+
+        @Override
+        public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
+            // The onFling() captures the 'up' event so our view thinks it gets long pressed. We don't want that, so cancel it.
+            if (odkView != null) {
+                odkView.cancelLongPress();
+            }
+            return false;
+        }
+
+        @Override
+        public void onLongPress(MotionEvent e) {
+
+        }
+
+        @Override
+        public boolean onSingleTapUp(MotionEvent e) {
+            return false;
+        }
+
+        @Override
+        public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
+            FlingRegister.flingDetected();
+
+            // only check the swipe if it's enabled in preferences
+            String navigation = (String) GeneralSharedPreferences.getInstance()
+                    .get(GeneralKeys.KEY_NAVIGATION);
+
+            if (e1 != null && e2 != null
+                    && navigation.contains(GeneralKeys.NAVIGATION_SWIPE) && allowSwiping) {
+                // Looks for user swipes. If the user has swiped, move to the appropriate screen.
+
+                // For all screens a swipe is left/right of at least .25" and up/down of less than .25" OR left/right of > .5"
+                int xpixellimit = (int) (ScreenUtils.getXPixelsNumberPerInch() * .25);
```

Comment:
I think this is clearer with `xdpi` and `ydpi` inline. The comment is very clear and `dpi` is common for "dots per inch".

Reply:
Right but still I think it's good to keep the code in a separate class (ScreenUtils) so I just change naming.

Reply:
I guess that by avoiding the direct dependency on the Collect instance it gets things a little closer to being testable. Fine by me.